key.go

```diff
@@ -691,22 +710,24 @@ func (k *Key) PrivateKey() (crypto.PrivateKey, error) {
 			curve = elliptic.P521()
 		}
 
-		priv := &ecdsa.PrivateKey{
-			PublicKey: ecdsa.PublicKey{Curve: curve, X: new(big.Int), Y: new(big.Int)},
-			D:         new(big.Int),
+		_, x, y, d := k.EC2()
+		var bx, by *big.Int
+		if len(x) == 0 || len(y) == 0 {
+			bx, by = curve.ScalarBaseMult(d)
+		} else {
+			bx = new(big.Int).SetBytes(x)
+			by = new(big.Int).SetBytes(y)
 		}
-		priv.X.SetBytes(x)
-		priv.Y.SetBytes(y)
-		priv.D.SetBytes(d)
+		bd := new(big.Int).SetBytes(d)
 
-		return priv, nil
+		return &ecdsa.PrivateKey{
+			PublicKey: ecdsa.PublicKey{Curve: curve, X: bx, Y: by},
+			D:         bd,
+		}, nil
 	case AlgorithmEd25519:
 		_, x, d := k.OKP()
-		// RFC8152 allows omitting X from private keys;
-		// crypto.PrivateKey assumes it is available.
-		// see https://www.rfc-editor.org/rfc/rfc8152#section-13.2
 		if len(x) == 0 {
-			return nil, ErrOKPNoPub
+			return ed25519.NewKeyFromSeed(d), nil
```

Comment:
I have a general question here: What does `d` mean for Ed25519 in terms of OKP? Is it private key or seed?

[RFC 8152 13.2](https://www.rfc-editor.org/rfc/rfc8152#section-13.2) says that
> d:   This contains the private key.

Reply:
d is a private key component of a crv, but my understanding of the specs is that it can be anything sufficiently random... so in the case that you derive a seed from a mnemonic or other details, you might use the seed as the private key... thats all super dangerous stuff... and I don't think any of it is relevant to COSE.

Reply:
The function name is a little misleading. Its [documentation](https://pkg.go.dev/crypto/ed25519#NewKeyFromSeed) is more interesting:
> NewKeyFromSeed calculates a private key from a seed. It will panic if len(seed) is not [SeedSize](https://pkg.go.dev/crypto/ed25519#SeedSize). This function is provided for interoperability with [RFC 8032](https://rfc-editor.org/rfc/rfc8032.html). [RFC 8032](https://rfc-editor.org/rfc/rfc8032.html)'s private keys correspond to seeds in this package.

[RFC8152 8.2](https://www.rfc-editor.org/rfc/rfc8152#section-8.2) says that EdDSA curves should be implemented following RFC8032, so my understanding is that it is safe to use `NewKeyFromSeed` here.